Add hasattr function for safe if attribute is not existed.

geonmo opened the pull request:
It is just a fix to avoid missing attribute error.

slehti committed:
Updated HLTTau validation to include legs X in tau+X triggers

slehti committed:
HLTTauDQML1Plotter: bugfix

ikrav committed:
New MVA IDs from Rong-Shyang for Spring15: V1 for 50ns and V0 for 25ns

ikrav committed:
Added the mvaTag field to the MVA estimator, plus dependencies, to distinguish different cases of the MVA of the same class configured with different weights/tunings.

ikrav committed:
Added the new mvaTag field to the most recent MVA configuratoins

skyriacoCMS committed:
Create effAreaNeutralHadrons_cone03_pf_50ns.txt

skyriacoCMS committed:
Create effAreaPhotons_cone03_pfPhotons_50ns.txt

skyriacoCMS committed:
Create effAreaPhotons_cone03_pfChargedHadrons_50ns.txt

skyriacoCMS committed:
Rename effAreaNeutralHadrons_cone03_pf_50ns.txt to effAreaPhotons_cone03_pfNeutralHadrons_50ns.txt

skyriacoCMS committed:
Create cutBasedPhotonID_Spring15_50ns_V1_cff.py

ikrav committed:
New effective areas for electrons for Spring15 25ns abd 50ns

ikrav committed:
The implementation of the new 25ns non-triggering MVA tuned by P.Pigard, the C++ part. The Python part is yet to come.

ikrav committed:
Renamed the two files of the new Spring15 electron non-trig MVA into the right file and class name

ikrav committed:
Weights and new python configuration for the new non-trig 25ns Spring15 electron MVA

ikrav committed:
A few fixes

ikrav committed:
a change to the Reader options

lgray committed:
update Spring15NonTrig for electrons to threadsafe base class

lgray committed:
Spring15 50ns ID from 31072015 presentation (Rami)

lgray committed:
Set Spring15_25ns MVA to actual values for WP90 from P. Pigard

lgray committed:
add 50ns mva Photon ID V2

lgray committed:
Fix implementation of @skyriacoCMS photon ID, fix bug in lower level python steering

lgray committed:
update MiniAOD tools to the latest EGM-blessed IDs

lgray committed:
invalidate deprecated IDs

lgray committed:
fix typos in Spring15 ID cutbased, validation now checks out

lgray committed:
remove HEEP51 from MiniAOD and deprecate it

lgray committed:
turn off embedding of vid::CutFlowResult in MiniAOD pat candidates

lgray committed:
EGM Object Modifiers for MiniAOD reprocessing

lgray committed:
add wp80 to Spring15 electron mva ID

lgray committed:
add in embedding control from 74X to 75X to vid tools

lgray committed:
convert ReducedEGProducer to edm::global

lgray committed:
convert PhysicsTools/PatAlgos and related to threaded modules

lgray committed:
convert TopProjector to edm::stream

lgray committed:
convert GenParticlePruner to edm::stream

lgray committed:
convert JetFlavourClustering to edm::stream

lgray committed:
convert JetDeltaRValueMapProducer to edm::global

lgray committed:
convert bits of RecoJets/JetProducers related to MiniAOD to threaded modules

lgray committed:
convert HadronAndPartonSelector to edm::stream

lgray committed:
convert JetChargeProducer to edm::global

lgray committed:
convert bits of JetMETCorrections/Type1MET and PhysicsTools/PatUtils related to MET and MiniAOD to threaded modules

lgray committed:
convert TauGenJetProducer to edm::global

lgray committed:
convert bits of CommonTools/RecoUtils related to MiniAOD to threaded modules

lgray committed:
changes to MethodInvoker for thread safety

lgray committed:
use constants instead of magic numbers eta cut

lgray committed:
fix semicolon;

lgray committed:
reduce copy/pasta in IsolationModifier

lgray committed:
fix small logic bug

lgray committed:
RecoTauCleaner, PFRecoTauTagInfoProducer, PFTauSecondaryVertexProducer to modern modules

lgray committed:
move citk producer to edm::stream

lgray committed:
convert used modules in RecoLocalCalo/HcalRecProducers to edm::stream

lgray committed:
convert L1TCaloUpgradeToGCTConverter to edm::global

lgray committed:
convert JetDeltaRTagInfoValueMapProducer to edm::global

ikrav committed:
Added python for cut-based photon MVA Spring15 25ns V2 and deprecated V0. Added this ID to the PhotonMVAValueMapProducer instead of V0

ikrav committed:
A new cut based electron ID, 25ns Spring15 first version

ikrav committed:
Fixed a typo in the eta ranges

lgray committed:
move NjettinessAdder to edm::one because of underlying statics

lgray committed:
code review resulting from helgrind, move modules from global to stream that contained reco::parser stuff

lgray committed:
update embedding to V2

lgray committed:
update IDs in MiniAOD config

lgray committed:
remove using directives from headers in PatAlgos/plugins

alja committed:
Update camera on clipping.

alja committed:
Destroy region clipped views on next event.

Conflicts:
	Fireworks/Core/src/FWGUIManager.cc

davidsheffield committed:
Add scouting electron, photon, and muon classes and producers

Martin-Grunewald committed:
Merged refs/pull/10836/head from repository cms-sw

Martin-Grunewald committed:
v5 L1T finally in CondDB

Martin-Grunewald committed:
Fix missing comma

perrotta committed:
Merged refs/pull/11043/head from repository cms-sw

perrotta committed:
Update the 25ns menu up to master /dev/CMSSW_7_4_0/HLT/V456

Commit:
Added a new path in DiPhoton category.

alja committed:
Rename clipping eta parameter to theta.

Conflicts:
	Fireworks/Core/src/FW3DViewBase.cc
	Fireworks/Core/src/FWGUIManager.cc

perrotta committed:
Update up to /dev/CMSSW_7_4_0/HLT/V456

slava77devel committed:
made up a default constructor

slava77devel committed:
made up a default constructor

slava77devel committed:
check element is valid before accessing

slava77devel committed:
check element is valid before accessing

slava77devel committed:
initialize array Pedestal

slava77devel committed:
initialize array Pedestal

slava77devel committed:
add missing semiE_LepJetPtCut parameter reading from config

slava77devel committed:
add missing semiE_LepJetPtCut parameter reading from config

Martin-Grunewald committed:
Merged 76X25nsHLT from repository cms-tsg-storm

Martin-Grunewald committed:
Fixed tag name for L1T

Martin-Grunewald committed:
Fixed tag name for L1T

geonmo committed:
Add hasattr function for safe if attribute is not existed.

cmsbuild commented:
A new Pull Request was created by @geonmo (Geonmo Ryu) for CMSSW_7_6_X.

Add hasattr function for safe if attribute is not existed.

It involves the following packages:

SLHCUpgradeSimulations/Configuration


@cmsbuild, @civanch, @mdhildreth can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@degano you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

geonmo committed:
More safe!

cmsbuild commented:
Pull request #11095 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 82442d7c4b7ec4722ae2e27d1cd626246579a0a4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11095/7846/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11095/7846/summary.html

slava77devel committed:
initialize arrays used for counters

slava77devel committed:
initialize arrays used for counters

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented on SLHCUpgradeSimulations/Configuration/python/gemCustom.py:
```diff
@@ -9,16 +9,17 @@ def customise2023(process):
-  process.simMuonRPCDigis.digiModel = cms.string('RPCSimAverageNoiseEff')
+  if ( not hasattr(process,"simMuonGEMDigis")) :
+    process = customize_digi_addGEM_muon_only(process)
+  if ( hasattr(process,"simMuonGEMDigis")) :
```
this can just be an "else", right?

geonmo committed:
Change if statement to avoid confusing. Add EDMtoME for our module.

cvuosalo committed:
Fixed conflicts

geonmo committed:
Remove wrong comment.

geonmo commented:
@davidlange6 I updated about if-statement. I think that "else" is not proper at this situation.

cmsbuild commented:
Pull request #11095 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

geonmo committed:
Remove duplecate path

cmsbuild commented:
Pull request #11095 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

Martin-Grunewald committed:
Dump of v4

davidlange6 commented on SLHCUpgradeSimulations/Configuration/python/gemCustom.py:
```diff
@@ -9,25 +9,24 @@ def customise2023(process):
-  process.simMuonRPCDigis.digiModel = cms.string('RPCSimAverageNoiseEff')
+  if ( not hasattr(process,"simMuonGEMDigis")) :
+    process = customize_digi_addGEM_muon_only(process)
+  if ( not hasattr(process,"simMuonGEMDigis.mixLabel")) :
```
@geonmo - I notice only now - this is not proper python. you need to test first hasattr(process,'simMuonGEMDigis') and then hasattr(process.simMuonGEMDigis,'mixLablel')

perrotta committed:
Merged 76X25nsHLT from repository cms-tsg-storm

perrotta committed:
Update lists for /dev/CMSSW_7_4_0/HLT/V462

Dr15Jones committed:
Made MaterialAccountingTrack const correct

MaterialAccountingTrack had exposed its member data as public and had
two mutable data members. This change puts the code in line with CMSSW
standards for data products.

slava77devel committed:
move initialization to the right place

slava77devel committed:
move initialization to the right place

davidsheffield committed:
Move producers to src directory

slava77devel committed:
maybe to fix an error of reading out of range

slava77devel committed:
maybe to fix an error of reading out of range

Dr15Jones committed:
Made PFTau const thread-safe

PFTau caches values the first time certain member functions are called.
The caches were changed to use edm::AtomicPtrCache in order to be
thread-safe.

Dr15Jones committed:
Use iorule to remove mutables from CandKinResolution

Previously, CandKinResolution would on the first request for the
covariance matrix fill the matrix. This was done since the matrix
is a transient member. This delayed filling is difficult to do in
a thread safe manner. Therefore the matrix is now filled when reading
back from the file via a ROOT iorule.

Dr15Jones committed:
Modifications to have friends work with new PFTau internals

matteosan1 committed:
Merged refs/pull/11118/head from repository cms-sw

matteosan1 committed:
roll-back of egammaObjectModification to solve VID issues

smorovic committed:
*bugfix:prevent premature exit when EoR file is found, in case when next LS is signalled to input source
*correct initialization of variable ls

perrotta committed:
Update up to /dev/CMSSW_7_4_0/HLT/V468

makortel committed:
Add function to convert string to RectangularEtaPhiTrackingRegion::UseMeasurementTracker enumeration

makortel committed:
Enclose sqr to anonymous namespace

makortel committed:
Fix MeasurementTrackerEvent configuration in L3MumuTrackingRegion

makortel committed:
Fix MeasurementTrackerEvent configuration in CandidateSeededTrackingRegionsProducer

makortel committed:
Fix MeasurementTrackerEvent configuration in TauRegionalPixelSeedGenerator

makortel committed:
Fix MeasurementTrackerEvent configuration in TrackingRegionsFromBeamSpotAndL2Tau

makortel committed:
Fix MeasurementTrackerEvent configuration in PointSeededTrackingRegionsProducer

slava77devel committed:
pass by const reference

slava77devel committed:
pass by const reference

perrotta committed:
Merged 76X25nsHLT from repository cms-tsg-storm

perrotta committed:
Remove wrongly added file

perrotta committed:
Update text files for subtables and script for 0T subtable

perrotta committed:
Update subtables

Martin-Grunewald committed:
Merged 76X25nsHLT from repository cms-tsg-storm

Martin-Grunewald committed:
frozen v4.1

Martin-Grunewald committed:
0T txt

Commit:
replaced logError+exit combinations with throw cms::Exception

alja committed:
Remove debug printout.

cvuosalo committed:
Merged refs/pull/11140/head from repository cms-sw

cvuosalo committed:
Fixed formatting

cmsbuild committed:
Merge pull request #11201 from lveldere/fastsim-noMoreExits

FastSim, use cms::Exception rather than logError + exit

davidlange6 committed:
Merge pull request #11173 from slava77/CMSSW_7_6_X_2015-08-30-2300/puMVAetc-valgrind

cleanup of uninitialized reads reported by valgrind

cmsbuild committed:
Merge pull request #11145 from Dr15Jones/constThreadSafePFTau

Made PFTau const thread-safe

cmsbuild committed:
Merge pull request #11064 from ndaci/ExoValDev_76Xto75X_PR5

Added a new path in DiPhoton category. (Backport from #11063)

Commit:
Updated paths (DiPhoton, HighPtPhoton).

Commit:
Update hltExoticaHighPtPhoton_cff.py

Add "0T" comment.

alja committed:
Fix clang warnings come back from failed merge.

davidlange6 committed:
Merge pull request #11175 from slava77/CMSSW_7_5_X_2015-09-06-2300/puMVAetc-valgrind

cleanup of uninitialized reads reported by valgrind (backport of #11173 )

davidlange6 committed:
Merge pull request #11160 from matteosan1/electronObjectModifier_75X_for_arizzi

Electron object modifier 75X

davidlange6 committed:
Merge pull request #11219 from ndaci/ExoValDev_76X_PR6_v2

 Updated paths (DiPhoton, HighPtPhoton).

cmsbuild committed:
Merge pull request #10967 from fwyzard/migrate_to_multithreading_75x_HLT

migrate modules used by the HLT menu to multithreading (HLT) (75x)

cmsbuild committed:
Merge pull request #10265 from slehti/HLTTauValidationUpdate75X_July2015

Hlt tau validation update75x

davidlange6 committed:
Merge pull request #11158 from smorovic/CMSSW_7_6_X-eorfix

Fix premature end of run in DAQ (76X)

davidlange6 committed:
Merge pull request #11146 from Dr15Jones/ioruleForCandKinResolution

Use iorule to remove mutables from CandKinResolution

davidlange6 committed:
Merge pull request #11140 from cvuosalo/merge-attempt

Convert remaining legacy modules in RECO to support multi-threading

davidlange6 commented:
ping @geonmo

cmsbuild committed:
Merge pull request #11217 from alja/75-clipcamera

75x Fireworks: Adjust camera position when open 3D region

cmsbuild committed:
Merge pull request #11183 from makortel/fixMTEModeInTrackingRegionProducers_v2

Fix MeasurementTrackerEvent configuration of some TrackingRegionProducers

cmsbuild committed:
Merge pull request #11043 from cms-tsg-storm/76X25nsHLT

76X 25ns HLT round three

cmsbuild committed:
Merge CMSSW_7_5_X into CMSSW_7_6_X.

geonmo committed:
Add mixlabel if parent collection is existed and mixlabel is missing.

geonmo commented:
Hello, @davidlange6 
I am sorry about late reply. I fix it as your comment.
Could you test?

cmsbuild commented:
Pull request #11095 was updated. @cmsbuild, @civanch, @mdhildreth can you please check and sign again.

davidlange6 committed:
Merge pull request #11135 from Dr15Jones/constCorrectMaterialAccountingTrack

Made MaterialAccountingTrack const correct

davidlange6 committed:
Merge pull request #11132 from davidsheffield/76X_ScoutingPacking_LeptonsPhoton

Add scouting electron, photon, and muon classes and producers

smuzaffar committed:
Merge pull request #11228 from smuzaffar/from-CMSSW_7_6_X_2015-09-09-2300

allowed oracle dependency for DQM/EcalMonitorDbModule

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e57de00a74a2fb38dff8ed86c964adbfd6bdb8e4
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11095/8029/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11095/8029/summary.html

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_6_X IBs  (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @degano, @smuzaffar

davidlange6 commented:
@geonmo - this does not fix at least one of my comments. Specifically

if ( not hasattr(process,"simMuonME0Digis.mixLabel")) 

will always return true. I don't think its the behavior that you want.

geonmo commented:
@davidlange6 Hello! Frankly speaking, I want to avoid missing "mixLabel" error about simMuonME0Digis.mixLabel due to default setting is wrong. It will be fixed but not now. So, I added this configuration on GEM Custom.

geonmo commented:
@davidlange6 If my code is correct, It will not be a problem even if simMuonME0Digis is not existed.

davidlange6 commented:
@geonmo - I'm not explaining well.. let me try again

hasattr(p,'x.y') is not valid python.

it will fail even if p.x.y exists.

geonmo committed:
Merged addHasAttr from repository geonmo

geonmo committed:
Fix the real bug at hasattr.

geonmo commented:
@davidlange6 Hello, I've got it your comment. I fix it properly. However, this PR is corructed any other commits from IB. I will re-submit a PR. Thanks.
